HW0/P6/P6.txt

```diff
@@ -1 +1,199 @@
+  example, 'Bye Job 0' appears after three other 'Bye Job' statements
+  have already been output. On the other hand, all 'Bye Job' statements
+  appear in order from 0 to 9 in run 4.
+
+- There is no fixed pattern of "Hi Job" and "Bye Job" statements; e.g.
+  it is not always the case that we'll see alternating "Hi Job" and 
+  "Bye Job" statements after a fixed number of "Hi Job" statements.
+
+- Although the 'sleep' time between Hi and Bye jobs of the same number
+  appear fixed, each pair is not separated by the same number of
+  statements.
+
+In general, the statement you expect to be output first for each category
+may not output first; the statement you expect to be output last for each
+category may not output last. Additionally, the ordering of both 
+the outputs within and between each category cannot be predicted with
+certainty in these example runs.
+
+If the programmer is not careful, operations may be performed in an
+order that was not intended. This could result in spurious or incorrect
+output, and may even crash programs, depending on the nature of the
+operations.
+
+For example, suppose that each "Hi Job" operation represented an
+operation to assign a variable one value, whereas a "Bye Job" operation
+might retrieve values for one or more variables and perform some
+mathematical operation on them. If the assignment and retrieval
+operations are not performed in the expected order, problems could
+occur: e.g. we could be retrieving the value for a variable to which
+an assignment has not yet been made, we could retrieve an old value
+rather than what was expected to be the current value (and use the
+wrong one in our mathematical operation, further perpetuating the
+miscalculation), we could retrieve a value that is too new (the one
+that we wanted has already been overwritten), and so on.
+
+Therefore, if certain operations must be performed in a specific
+sequence relative to other operations in order to produce correct
+program behavior, we as parallel programmers need a mechanism to enforce
+the order for these particular operations and prevent them from being
+changed. In this regard, when multiple processes are involved, and
+when one process wants to access the same resource (e.g. memory) as
+another contemporaneously, we need a mechanism to control access to 
+those resources to prevent one process from intervening with another's
+access in unexpected and potentially harmful ways.
+
+Another example scenario involves the calculations we performed in
+Problem 4, for columns in matrix S as an example. Each set of values for
+a given column in s depended on values of the preceding column (which is
+associated with the previous time step); if we try to apply parallel
+programming techniques and do so in a way that changes the order of the
+columns for example, the resultant model will be different (and likely
+incorrect).
+
+2. Running P6B.py
+
+From the plot, in general there seems to be a trend that suggests the
```

Comment:
I pretty agree with your answer. Under the condition from 10^(-6)—>10^(-4), since the task is relatively small, serial version is faster than parallel version. However, under the condition from 10^(-3)->10^(0), the task is relatively large, parallel version will take shorter than serial version. Therefore, if we program this in parallel, it will take much more time to run this program if the task is relatively small, since opening a processor costs much more time. But if the task is relatively large, using multiprocessor will make the whole process much faster.